lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -4959,6 +4959,33 @@ void TypeChecker::checkConformancesInContext(DeclContext *dc,
         diagnoseMissingAppendInterpolationMethod(*this, typeDecl);
       }
     }
+      
+    if (conformance->getProtocol()->
+          isSpecificProtocol(KnownProtocolKind::OptionSet)) {
+          for (auto member : idc->getMembers()) {
+            if (auto pbd = dyn_cast<PatternBindingDecl>(member)) {
+              if (pbd->isStatic())
+                for (auto entry : pbd->getPatternList()) {
+                  if (auto ctor = dyn_cast<CallExpr>(entry.getInit())) {
+                    auto type = ctor->getType();
+                    auto argLabels = ctor->getArgumentLabels();
+                    auto *args = cast<TupleExpr>(ctor->getArg());
+                    if (type->isEqual(dc->getSelfTypeInContext())) {
+                      if (argLabels[0].is(StringRef("rawValue"))) {
+                        if (auto i = dyn_cast<IntegerLiteralExpr>(args->getElement(0))) {
+                          auto a = i->getValue();
+                          if (a == 0) {
+                            auto loc = member->getLoc();
+                            diagnose(loc, diag::option_set_zero_constant, type);
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+          }
+      }
```

Comment:
If you're looking for advice how to make this more readable, the two biggest recommendations are
- Split it into a function, and
- Early-return

Something like this could help, with comments explaining the checks you're making:

```c++
void diagnoseEmptyOptionSetCase(/* args you need... */) {
  auto protocol = conformance->getProtocol();
  if (!protocol->isSpecificProtocol(KnownProtocolKind::OptionSet))
    return;

  for (auto member : idc->getMembers()) {
    auto pbd = dyn_cast<PatternBindingDecl>(member);
    if (!pbd || !pbd->isStatic()) continue;
    // ... rest of the stuff, but early-returning/continuing when the condition is not met
  }
}
```

Reply:
Or take Jordan's suggestion, which should simplify the implementation a bit 😄

Reply:
Ah, thanks!

Reply:
Would you be able to point me in right direction in the codebase? Would that be done at parsing time or at type checking?

Reply:
This would happen during type checking, after we've assigned types to stored properties. `visitBoundVariable` in `TypeCheckDecl.cpp` might be a good place to put this check.

Reply:
Ok thanks!